### PR TITLE
Migrate from PyPI tokens to Trusted Publishers

### PR DESCRIPTION
**Pull request by 0x2b3bfa0**
https://blog.pypi.org/posts/2023-04-20-introducing-trusted-publishers#using-trusted-publishing-with-github-actions

### CODE REVIEW

**Comment by 0x2b3bfa0** [comment]
Irrelevant, semi-automated change, for consistency with the other repositories.

**Comment by skshetry** [comment]
@0x2b3bfa0, we are already using trusted-publishers. I don't think `environment` adds anything for us, unless you can convince me otherwise.

**Comment by 0x2b3bfa0** [comment]
I know, I know. 😄

I was just updating every single PyPI release workflow as part of https://github.com/iterative/itops/issues/2499, so all of them look similar.

Environments would allow you to track better the status of deployments and restrict publishing packages to a narrower set of users than those with write access to the repository, if you ever need to.

Still, if you feel strongly about this, feel free to close the pull request unmerged; this change is technically irrelevant.

File: .github/workflows/build.yaml
Comment (0x2b3bfa0) [inline .github/workflows/build.yaml:12]:
```suggestion
```

@skshetry, feel free to apply this change before merging.

File: .github/workflows/build.yaml
Comment (skshetry) [inline .github/workflows/build.yaml:15]: Could you please explain this change please?

Comment (0x2b3bfa0) [reply .github/workflows/build.yaml:15]: > [_only set the `id-token: write` permission in the job that does publishing, not globally_](https://github.com/pypa/gh-action-pypi-publish/blob/a712d989cca7b5d2b239f655efb2512082d5bce8/README.md?plain=1#L81-L86)

Albeit it's arguably pointless in a single-job workflow, the recommendation is to set the permissions at the job level rather than globally.

TL;DR another unnecessary change you may refuse within reason.

Comment (skshetry) [reply .github/workflows/build.yaml:14]: Are those permissions merged? If yes, then I'd prefer `contents: read` to be on the top-level and `id-write` on jobs level. Otherwise, I'm okay with your changes. :)

Comment (0x2b3bfa0) [reply .github/workflows/build.yaml:14]: > _Are those permissions merged?_

They aren't, unfortunately. The `permissions` block on jobs overrides the global one.

Comment (skshetry) [reply .github/workflows/build.yaml:17]: Okay, I added `contents: read` back at the top-level so that it'll be used if we add additional jobs here.

**Review by skshetry — APPROVED** [review]

**Comment by skshetry** [comment]
Thank you @0x2b3bfa0.

**Pull request merged**
